File: src/go/radio/api_radio.go
package radio

import "time"

type RadioApi struct {
	*ApiModule
}

func NewRadioApi(rm *RadioManager) *RadioApi {
	return &RadioApi{NewApiModule(rm)}
}

func (api *RadioApi) IsSceneInit() interface{} {
	return api.Send("radio.isSceneInit", nil, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["init"]
	})
}

func (api *RadioApi) IsRadioReachabilityCheckerInit() interface{} {
	return api.Send("radio.isRadioReachabilityCheckerInit", nil, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["init"]
	})
}

func (api *RadioApi) CheckReachability(aTx XYZ, bRx XYZ, options *RadioCheckOptions) interface{} {
	data := map[string]interface{}{
		"aTx":     []float64{aTx[0], aTx[1], aTx[2]},
		"bRx":     []float64{bRx[0], bRx[1], bRx[2]},
		"options": options,
	}
	return api.Send("radio.checkReachability", data, "", 3*time.Second, nil)
}

func (api *RadioApi) UpdateObjectPos(objectId string, position XYZ) interface{} {
	data := map[string]interface{}{
		"objectId": objectId,
		"position": []float64{position[0], position[1], position[2]},
	}
	return api.Send("radio.updateObjectPos", data, "", 3*time.Second, nil)
}

func (api *RadioApi) GetObjectPos(objectId string) interface{} {
	data := map[string]interface{}{
		"objectId": objectId,
	}
	return api.Send("radio.updateObjectPos", data, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["position"]
	})
}

func (api *RadioApi) UpdateMeshRadioMaterial(meshId string, materialId *string, thicknessM *float64) interface{} {
	data := map[string]interface{}{
		"meshId":      meshId,
		"materialId":  materialId,
		"thickness_m": thicknessM,
	}
	return api.Send("radio.updateMeshRadioMaterial", data, "", 3*time.Second, nil)
}

func (api *RadioApi) GetAllRadioMaterial() interface{} {
	return api.Send("radio.getAllRadioMaterial", nil, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["meshIds"]
	})
}

func (api *RadioApi) LocalRadioMaterial() interface{} {
	return api.Send("radio.localRadioMaterial", nil, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["meshIds"]
	})
}

func (api *RadioApi) GetBuildingRadioMaterial() interface{} {
	return api.Send("radio.getBuildingRadioMaterial", nil, "", 3*time.Second, nil)
}

func (api *RadioApi) AddRadioMaterial(material RadioMaterialProperties) interface{} {
	data := map[string]interface{}{
		"id":                         material.ID,
		"displayName":                material.DisplayName,
		"penetrationLoss_dBPerMeter": material.PenetrationLoss_dBPerMeter,
		"reflectionCoefficient":      material.ReflectionCoefficient,
		"defaultThickness_m":         material.DefaultThicknessM,
	}
	return api.Send("radio.addRadioMaterial", data, "", 3*time.Second, nil)
}

func (api *RadioApi) ListRadioLocalObjectsIds() interface{} {
	return api.Send("radio.listRadioLocalObjects", nil, "", 3*time.Second, func(d map[string]interface{}) interface{} {
		return d["localObjectIds"]
	})
}

File: src/rs/src/radio/api_radio.rs
use crate::radio::api_module::{ApiModule, SendResult};
use crate::radio::type_def::{XYZ, RadioCheckOptions, RadioMaterialProperties};
use serde_json::json;

pub struct RadioApi {
    pub base: ApiModule,
}

impl RadioApi {
    pub fn new(base: ApiModule) -> Self {
        Self { base }
    }

    pub fn is_scene_init(&self) -> SendResult {
        // 在 Rust 中，post_processor 需要在 RadioManager 层处理或在此处处理结果
        self.base.send("radio.isSceneInit", None, None)
    }

    pub fn is_radio_reachability_checker_init(&self) -> SendResult {
        self.base.send("radio.isRadioReachabilityCheckerInit", None, None)
    }

    pub fn check_reachability(&self, a_tx: XYZ, b_rx: XYZ, options: Option<RadioCheckOptions>) -> SendResult {
        self.base.send("radio.checkReachability", Some(json!({
            "aTx": a_tx,
            "bRx": b_rx,
            "options": options
        })), None)
    }

    pub fn update_object_pos(&self, object_id: &str, position: XYZ) -> SendResult {
        self.base.send("radio.updateObjectPos", Some(json!({
            "objectId": object_id,
            "position": position
        })), None)
    }

    pub fn get_object_pos(&self, object_id: &str) -> SendResult {
        self.base.send("radio.updateObjectPos", Some(json!({
            "objectId": object_id
        })), None)
    }

    pub fn update_mesh_radio_material(&self, mesh_id: &str, material_id: Option<String>, thickness_m: Option<f64>) -> SendResult {
        self.base.send("radio.updateMeshRadioMaterial", Some(json!({
            "meshId": mesh_id,
            "materialId": material_id,
            "thickness_m": thickness_m
        })), None)
    }

    pub fn get_all_radio_material(&self) -> SendResult {
        self.base.send("radio.getAllRadioMaterial", None, None)
    }

    pub fn local_radio_material(&self) -> SendResult {
        self.base.send("radio.localRadioMaterial", None, None)
    }

    pub fn get_building_radio_material(&self) -> SendResult {
        self.base.send("radio.getBuildingRadioMaterial", None, None)
    }

    pub fn add_radio_material(&self, material: RadioMaterialProperties) -> SendResult {
        self.base.send("radio.addRadioMaterial", Some(json!(material)), None)
    }

    pub fn list_radio_local_objects_ids(&self) -> SendResult {
        self.base.send("radio.listRadioLocalObjects", None, None)
    }
}

impl Default for RadioApi {
    fn default() -> Self {
        Self { base: ApiModule::default() }
    }
}
